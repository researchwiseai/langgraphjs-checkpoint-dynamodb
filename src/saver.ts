import { DynamoDBClient, type DynamoDBClientConfig } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb';
import type { RunnableConfig } from '@langchain/core/runnables';
import {
    BaseCheckpointSaver,
    type SerializerProtocol,
    type CheckpointTuple,
    type Checkpoint,
    type CheckpointMetadata,
    type CheckpointPendingWrite,
    type CheckpointListOptions,
    type PendingWrite,
} from '@langchain/langgraph-checkpoint';
import type { ValidatedConfigurable, CheckpointItem } from './types';
import { DynamoDBWriteItem, Write } from './write';

/**
 * DynamoDBSaver is a class that provides persistence to
 * Langgraph's graphs using AWS's DynamoDB.
 *
 * @class
 * @extends BaseCheckpointSaver
 *
 * @param {Object} params - The parameters for the constructor.
 * @param {DynamoDBClientConfig} [params.clientConfig] - Optional configuration for the DynamoDB client.
 * @param {SerializerProtocol} [params.serde] - Optional serializer protocol for serializing and deserializing data.
 * @param {string} params.checkpointsTableName - The name of the DynamoDB table for storing checkpoints.
 * @param {string} params.writesTableName - The name of the DynamoDB table for storing writes.
 *
 * @property {DynamoDBClient} client - The DynamoDB client instance.
 * @property {DynamoDBDocument} docClient - The DynamoDB document client instance.
 * @property {string} checkpointsTableName - The name of the DynamoDB table for storing checkpoints.
 * @property {string} writesTableName - The name of the DynamoDB table for storing writes.
 *
 * @method getTuple - Retrieves a checkpoint tuple based on the provided configuration.
 * @param {RunnableConfig} config - The configuration for the runnable.
 * @returns {Promise<CheckpointTuple | undefined>} - A promise that resolves to a checkpoint tuple or undefined.
 *
 * @method list - Lists checkpoint tuples based on the provided configuration and options.
 * @param {RunnableConfig} config - The configuration for the runnable.
 * @param {CheckpointListOptions} [options] - Optional options for listing checkpoints.
 * @returns {AsyncGenerator<CheckpointTuple>} - An async generator that yields checkpoint tuples.
 *
 * @method put - Saves a checkpoint and its metadata to the DynamoDB table.
 * @param {RunnableConfig} config - The configuration for the runnable.
 * @param {Checkpoint} checkpoint - The checkpoint to save.
 * @param {CheckpointMetadata} metadata - The metadata associated with the checkpoint.
 * @returns {Promise<RunnableConfig>} - A promise that resolves to the updated runnable configuration.
 *
 * @method putWrites - Saves pending writes to the DynamoDB table.
 * @param {RunnableConfig} config - The configuration for the runnable.
 * @param {PendingWrite[]} writes - The pending writes to save.
 * @param {string} taskId - The task ID associated with the writes.
 * @returns {Promise<void>} - A promise that resolves when the writes are saved.
 *
 * @private
 * @method getWritePartitionKey - Generates a partition key for a write item.
 * @param {Object} item - The write item.
 * @param {string} item.thread_id - The thread ID.
 * @param {string} item.checkpoint_id - The checkpoint ID.
 * @param {string} item.checkpoint_ns - The checkpoint namespace.
 * @returns {string} - The generated partition key.
 *
 * @private
 * @method getWriteSortKey - Generates a sort key for a write item.
 * @param {Object} item - The write item.
 * @param {string} item.task_id - The task ID.
 * @param {number} item.idx - The index of the write.
 * @returns {string} - The generated sort key.
 *
 * @private
 * @method validateConfigurable - Validates the configurable object.
 * @param {Record<string, unknown> | undefined} configurable - The configurable object to validate.
 * @returns {ValidatedConfigurable} - The validated configurable object.
 * @throws {Error} - Throws an error if the configurable object is invalid.
 */
export class DynamoDBSaver extends BaseCheckpointSaver {
    private client: DynamoDBClient;
    private docClient: DynamoDBDocument;
    private checkpointsTableName: string;
    private writesTableName: string;

    constructor({
        clientConfig,
        serde,
        checkpointsTableName,
        writesTableName,
    }: {
        clientConfig?: DynamoDBClientConfig;
        serde?: SerializerProtocol;
        checkpointsTableName: string;
        writesTableName: string;
    }) {
        super(serde);
        this.client = new DynamoDBClient(clientConfig || {});
        this.docClient = DynamoDBDocument.from(this.client);
        this.checkpointsTableName = checkpointsTableName;
        this.writesTableName = writesTableName;
    }

    async getTuple(config: RunnableConfig): Promise<CheckpointTuple | undefined> {
        const getItem = async (configurable: ValidatedConfigurable) => {
            if (configurable.checkpoint_id != null) {
                // Use get
                const item = await this.docClient.get({
                    TableName: this.checkpointsTableName,
                    Key: {
                        thread_id: configurable.thread_id,
                        checkpoint_id: configurable.checkpoint_id,
                    },
                });

                return item.Item as CheckpointItem | undefined;
            } else {
                // Use query
                const result = await this.docClient.query({
                    TableName: this.checkpointsTableName,
                    KeyConditionExpression: 'thread_id = :thread_id',
                    ExpressionAttributeValues: {
                        ':thread_id': configurable.thread_id,
                    },
                    ...(configurable.checkpoint_ns && {
                        FilterExpression: `checkpoint_ns = ${configurable.checkpoint_ns}`,
                    }),
                    Limit: 1,
                    ConsistentRead: true,
                    ScanIndexForward: false, // Descending order
                });

                return result.Items?.[0] as CheckpointItem | undefined;
            }
        };

        const item = await getItem(this.validateConfigurable(config.configurable));
        if (!item) {
            return undefined;
        }

        const checkpoint = (await this.serde.loadsTyped(item.type, item.checkpoint)) as Checkpoint;
        const metadata = (await this.serde.loadsTyped(
            item.type,
            item.metadata
        )) as CheckpointMetadata;

        // Fetch pending writes
        const writesResult = await this.docClient.query({
            TableName: this.writesTableName,
            KeyConditionExpression:
                'thread_id_checkpoint_id_checkpoint_ns = :thread_id_checkpoint_id_checkpoint_ns',
            ExpressionAttributeValues: {
                ':thread_id_checkpoint_id_checkpoint_ns': Write.getPartitionKey(item),
            },
        });

        const pendingWrites: CheckpointPendingWrite[] = [];
        if (writesResult.Items) {
            for (const writeItem of writesResult.Items as DynamoDBWriteItem[]) {
                const write = Write.fromDynamoDBItem(writeItem);
                const value = await this.serde.loadsTyped(write.type, write.value);
                pendingWrites.push([write.task_id, write.channel, value]);
            }
        }

        return {
            config: {
                configurable: {
                    thread_id: item.thread_id,
                    checkpoint_ns: item.checkpoint_ns,
                    checkpoint_id: item.checkpoint_id,
                },
            },
            checkpoint,
            metadata,
            parentConfig: item.parent_checkpoint_id
                ? {
                      configurable: {
                          thread_id: item.thread_id,
                          checkpoint_ns: item.checkpoint_ns,
                          checkpoint_id: item.parent_checkpoint_id,
                      },
                  }
                : undefined,
            pendingWrites,
        };
    }

    async *list(
        config: RunnableConfig,
        options?: CheckpointListOptions
    ): AsyncGenerator<CheckpointTuple> {
        const { limit, before } = options ?? {};
        const thread_id = config.configurable?.thread_id;

        const expressionAttributeValues: Record<string, unknown> = {
            ':thread_id': thread_id,
        };
        let keyConditionExpression = 'thread_id = :thread_id';

        if (before?.configurable?.checkpoint_id) {
            keyConditionExpression += ' AND checkpoint_id < :before_checkpoint_id';
            expressionAttributeValues[':before_checkpoint_id'] = before.configurable.checkpoint_id;
        }

        const result = await this.docClient.query({
            TableName: this.checkpointsTableName,
            KeyConditionExpression: keyConditionExpression,
            ExpressionAttributeValues: expressionAttributeValues,
            Limit: limit,
            ScanIndexForward: false, // Descending order
        });

        if (result.Items) {
            for (const item of result.Items as CheckpointItem[]) {
                const checkpoint = (await this.serde.loadsTyped(
                    item.type,
                    item.checkpoint
                )) as Checkpoint;
                const metadata = (await this.serde.loadsTyped(
                    item.type,
                    item.metadata
                )) as CheckpointMetadata;

                yield {
                    config: {
                        configurable: {
                            thread_id: item.thread_id,
                            checkpoint_ns: item.checkpoint_ns,
                            checkpoint_id: item.checkpoint_id,
                        },
                    },
                    checkpoint,
                    metadata,
                    parentConfig: item.parent_checkpoint_id
                        ? {
                              configurable: {
                                  thread_id: item.thread_id,
                                  checkpoint_ns: item.checkpoint_ns,
                                  checkpoint_id: item.parent_checkpoint_id,
                              },
                          }
                        : undefined,
                };
            }
        }
    }

    async put(
        config: RunnableConfig,
        checkpoint: Checkpoint,
        metadata: CheckpointMetadata
    ): Promise<RunnableConfig> {
        const { thread_id } = this.validateConfigurable(config.configurable);

        const [type1, serializedCheckpoint] = this.serde.dumpsTyped(checkpoint);
        const [type2, serializedMetadata] = this.serde.dumpsTyped(metadata);

        if (type1 !== type2) {
            throw new Error('Failed to serialize checkpoint and metadata to the same type.');
        }

        const item: CheckpointItem = {
            thread_id,
            checkpoint_ns: config.configurable?.checkpoint_ns ?? '',
            checkpoint_id: checkpoint.id!,
            parent_checkpoint_id: config.configurable?.checkpoint_id,
            type: type1,
            checkpoint: serializedCheckpoint,
            metadata: serializedMetadata,
        };

        await this.docClient.put({
            TableName: this.checkpointsTableName,
            Item: item,
        });

        return {
            configurable: {
                thread_id: item.thread_id,
                checkpoint_ns: item.checkpoint_ns,
                checkpoint_id: item.checkpoint_id,
            },
        };
    }

    async putWrites(config: RunnableConfig, writes: PendingWrite[], taskId: string): Promise<void> {
        const { thread_id, checkpoint_ns, checkpoint_id } = this.validateConfigurable(
            config.configurable
        );

        if (checkpoint_id == null) {
            throw new Error('Missing checkpoint_id');
        }

        const writeItems = writes.map((write, idx) => {
            const [type, serializedValue] = this.serde.dumpsTyped(write[1]);
            const item = new Write({
                thread_id,
                checkpoint_ns,
                checkpoint_id,
                task_id: taskId,
                idx,
                channel: write[0],
                type,
                value: serializedValue,
            });

            return {
                PutRequest: {
                    Item: item.toDynamoDBItem(),
                },
            };
        });

        // Batch write items
        const batches = [];
        for (let i = 0; i < writeItems.length; i += 25) {
            batches.push(writeItems.slice(i, i + 25));
        }

        for (const batch of batches) {
            await this.docClient.batchWrite({
                RequestItems: {
                    [this.writesTableName]: batch,
                },
            });
        }
    }

    private getWritePartitionKey(item: {
        thread_id: string;
        checkpoint_id: string;
        checkpoint_ns: string;
    }): string {
        return `${item.thread_id}:${item.checkpoint_id}:${item.checkpoint_ns}`;
    }

    private getWriteSortKey(item: { task_id: string; idx: number }): string {
        return `${item.task_id}:${item.idx}`;
    }

    private validateConfigurable(
        configurable: Record<string, unknown> | undefined
    ): ValidatedConfigurable {
        if (!configurable) {
            throw new Error('Missing configurable');
        }

        const { thread_id, checkpoint_ns, checkpoint_id } = configurable;

        if (typeof thread_id !== 'string') {
            throw new Error('Invalid thread_id');
        }

        if (typeof checkpoint_ns !== 'string' && checkpoint_ns !== undefined) {
            throw new Error('Invalid checkpoint_ns');
        }

        if (typeof checkpoint_id !== 'string' && checkpoint_id !== undefined) {
            throw new Error('Invalid checkpoint_id');
        }

        return {
            thread_id,
            checkpoint_ns: checkpoint_ns ?? '',
            checkpoint_id: checkpoint_id,
        };
    }
}
