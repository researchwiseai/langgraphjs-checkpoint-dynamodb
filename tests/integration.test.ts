import { describe, it, expect, beforeEach, afterEach, setSystemTime } from 'bun:test';
import { DynamoDBSaver } from '../src/saver';
import {
    DynamoDBClient,
    CreateTableCommand,
    DeleteTableCommand,
    DescribeTableCommand,
} from '@aws-sdk/client-dynamodb';
import { CheckpointMetadata, uuid6 } from '@langchain/langgraph-checkpoint';
import { Annotation, END, START, StateGraph } from '@langchain/langgraph';
import { AIMessage, BaseMessage, HumanMessage } from '@langchain/core/messages';

setSystemTime(new Date('2022-01-01T03:00:00.000Z'));

// Helper function to wait for table to become ACTIVE
async function waitForTableActive(client: DynamoDBClient, tableName: string) {
    while (true) {
        const { Table } = await client.send(new DescribeTableCommand({ TableName: tableName }));
        if (Table?.TableStatus === 'ACTIVE') {
            break;
        }
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
}

async function waitForTableDeleted(client: DynamoDBClient, tableName: string) {
    while (true) {
        try {
            await client.send(new DescribeTableCommand({ TableName: tableName }));
        } catch (e) {
            if (e.name === 'ResourceNotFoundException') {
                break;
            }
        }
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
}

describe('DynamoDBSaver', () => {
    const checkpointsTableName = 'checkpoints';
    const writesTableName = 'writes';

    const saver = new DynamoDBSaver({
        clientConfig: {
            endpoint: process.env.AWS_DYNAMODB_ENDPOINT,
        },
        checkpointsTableName,
        writesTableName,
    });

    describe('integration with DynamoDB', () => {
        beforeEach(async () => {
            console.log('Creating tables');

            const client = new DynamoDBClient({
                endpoint: process.env.AWS_DYNAMODB_ENDPOINT,
            });

            await client.send(
                new CreateTableCommand({
                    TableName: checkpointsTableName,
                    KeySchema: [
                        { AttributeName: 'thread_id', KeyType: 'HASH' }, // Partition key
                        { AttributeName: 'checkpoint_id', KeyType: 'RANGE' }, // Sort key
                    ],
                    AttributeDefinitions: [
                        { AttributeName: 'thread_id', AttributeType: 'S' },
                        { AttributeName: 'checkpoint_id', AttributeType: 'S' },
                    ],
                    BillingMode: 'PAY_PER_REQUEST',
                })
            );

            await client.send(
                new CreateTableCommand({
                    TableName: writesTableName,
                    KeySchema: [
                        { AttributeName: 'thread_id_checkpoint_id_checkpoint_ns', KeyType: 'HASH' }, // Partition key
                        { AttributeName: 'task_id_idx', KeyType: 'RANGE' }, // Sort key
                    ],
                    AttributeDefinitions: [
                        {
                            AttributeName: 'thread_id_checkpoint_id_checkpoint_ns',
                            AttributeType: 'S',
                        },
                        { AttributeName: 'task_id_idx', AttributeType: 'S' },
                    ],
                    BillingMode: 'PAY_PER_REQUEST',
                })
            );

            await waitForTableActive(client, checkpointsTableName);
            await waitForTableActive(client, writesTableName);

            console.log('Tables created');
        });

        afterEach(async () => {
            console.log('Deleting tables');
            const client = new DynamoDBClient({
                endpoint: process.env.AWS_DYNAMODB_ENDPOINT,
            });

            await client.send(
                new DeleteTableCommand({
                    TableName: checkpointsTableName,
                })
            );

            await client.send(
                new DeleteTableCommand({
                    TableName: writesTableName,
                })
            );

            await waitForTableDeleted(client, checkpointsTableName);
            await waitForTableDeleted(client, writesTableName);

            console.log('Tables deleted');
        });

        it('should save and load checkpoints', async () => {
            const checkpoint = {
                v: 1,
                id: uuid6(-1),
                ts: '2024-04-19T17:19:07.952Z',
                channel_values: {
                    someKey1: 'someValue1',
                },
                channel_versions: {
                    someKey2: 1,
                },
                versions_seen: {
                    someKey3: {
                        someKey4: 1,
                    },
                },
                pending_sends: [],
            };

            await saver.put({ configurable: { thread_id: '1' } }, checkpoint, {
                source: 'update',
                step: -1,
                writes: null,
            } as CheckpointMetadata);

            const loadedCheckpoint = await saver.getTuple({
                configurable: { thread_id: '1' },
            });

            expect(loadedCheckpoint).not.toBeUndefined();
            expect(loadedCheckpoint?.checkpoint.id).toEqual(checkpoint.id);
        });

        it('should save and load writes', async () => {
            const checkpoint = {
                v: 1,
                id: uuid6(-1),
                ts: '2024-04-19T17:19:07.952Z',
                channel_values: {
                    someKey1: 'someValue1',
                },
                channel_versions: {
                    someKey2: 1,
                },
                versions_seen: {
                    someKey3: {
                        someKey4: 1,
                    },
                },
                pending_sends: [],
            };

            const writes = {
                writes: [
                    {
                        id: '1',
                        v: 1,
                        ts: '2024-04-19T17:19:07.952Z',
                        channel_values: {
                            someKey1: 'someValue1',
                        },
                        channel_versions: {
                            someKey2: 1,
                        },
                        versions_seen: {
                            someKey3: {
                                someKey4: 1,
                            },
                        },
                        pending_sends: [],
                    },
                ],
            };

            await saver.put({ configurable: { thread_id: '1' } }, checkpoint, {
                source: 'update',
                step: -1,
                writes,
                parents: {},
            } as CheckpointMetadata);

            const loadedWrites = await saver.getTuple({
                configurable: { thread_id: '1' },
            });

            expect(loadedWrites).not.toBeUndefined();
            expect(loadedWrites?.metadata?.writes).toEqual(writes);
        });

        describe('and a workflow', () => {
            it('should save and load history', async () => {
                const AgentState = Annotation.Root({
                    messages: Annotation<BaseMessage[]>({
                        reducer: (x, y) => x.concat(y),
                        default: () => [],
                    }),
                });

                const workflow = new StateGraph(AgentState)
                    .addNode('NodeA', async () => {
                        return {
                            messages: [
                                new AIMessage({
                                    content: 'Hello from NodeA',
                                }),
                            ],
                        };
                    })
                    .addNode('NodeB', async () => {
                        return {
                            messages: [
                                new AIMessage({
                                    content: 'Hello from NodeB',
                                }),
                            ],
                        };
                    });

                workflow.addEdge(START, 'NodeA');
                workflow.addEdge('NodeA', 'NodeB');
                workflow.addEdge('NodeB', END);

                const graph = workflow.compile({
                    checkpointer: saver,
                });

                const config = { configurable: { thread_id: '1' } };

                let loadedTuple = await saver.getTuple(config);
                expect(loadedTuple).toBeUndefined();

                const answer = await graph.invoke(
                    {
                        messages: [
                            new HumanMessage({
                                content: 'Hello from Human',
                            }),
                        ],
                    },
                    config
                );

                expect(answer).toMatchSnapshot();

                loadedTuple = await saver.getTuple(config);

                expect(loadedTuple).toMatchSnapshot();
            });
        });
    });
});
