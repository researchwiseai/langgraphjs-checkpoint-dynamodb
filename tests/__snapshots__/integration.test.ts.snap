// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamoDBSaver integration with DynamoDB and a workflow should save and load history 1`] = `
{
  "messages": [
    HumanMessage HumanMessage {
      "additional_kwargs": {},
      "content": "Hello from Human",
      "id": undefined,
      "lc_kwargs": {
        "additional_kwargs": {},
        "content": "Hello from Human",
        "response_metadata": {},
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": undefined,
      "response_metadata": {},
    },
    AIMessage AIMessage {
      "additional_kwargs": {},
      "content": "Hello from NodeA",
      "id": undefined,
      "invalid_tool_calls": [],
      "lc_kwargs": {
        "additional_kwargs": {},
        "content": "Hello from NodeA",
        "invalid_tool_calls": [],
        "response_metadata": {},
        "tool_calls": [],
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": undefined,
      "response_metadata": {},
      "tool_calls": [],
      "usage_metadata": undefined,
    },
    AIMessage AIMessage {
      "additional_kwargs": {},
      "content": "Hello from NodeB",
      "id": undefined,
      "invalid_tool_calls": [],
      "lc_kwargs": {
        "additional_kwargs": {},
        "content": "Hello from NodeB",
        "invalid_tool_calls": [],
        "response_metadata": {},
        "tool_calls": [],
      },
      "lc_namespace": [
        "langchain_core",
        "messages",
      ],
      "lc_serializable": true,
      "name": undefined,
      "response_metadata": {},
      "tool_calls": [],
      "usage_metadata": undefined,
    },
  ],
}
`;

exports[`DynamoDBSaver integration with DynamoDB and a workflow should save and load history 2`] = `
{
  "checkpoint": {
    "channel_values": {
      "NodeB": "NodeB",
      "messages": [
        HumanMessage HumanMessage {
          "additional_kwargs": {},
          "content": "Hello from Human",
          "id": undefined,
          "lc_kwargs": {
            "additional_kwargs": {},
            "content": "Hello from Human",
            "response_metadata": {},
          },
          "lc_namespace": [
            "langchain_core",
            "messages",
          ],
          "lc_serializable": true,
          "name": undefined,
          "response_metadata": {},
        },
        AIMessage AIMessage {
          "additional_kwargs": {},
          "content": "Hello from NodeA",
          "id": undefined,
          "invalid_tool_calls": [],
          "lc_kwargs": {
            "additional_kwargs": {},
            "content": "Hello from NodeA",
            "invalid_tool_calls": [],
            "response_metadata": {},
            "tool_calls": [],
          },
          "lc_namespace": [
            "langchain_core",
            "messages",
          ],
          "lc_serializable": true,
          "name": undefined,
          "response_metadata": {},
          "tool_calls": [],
          "usage_metadata": undefined,
        },
        AIMessage AIMessage {
          "additional_kwargs": {},
          "content": "Hello from NodeB",
          "id": undefined,
          "invalid_tool_calls": [],
          "lc_kwargs": {
            "additional_kwargs": {},
            "content": "Hello from NodeB",
            "invalid_tool_calls": [],
            "response_metadata": {},
            "tool_calls": [],
          },
          "lc_namespace": [
            "langchain_core",
            "messages",
          ],
          "lc_serializable": true,
          "name": undefined,
          "response_metadata": {},
          "tool_calls": [],
          "usage_metadata": undefined,
        },
      ],
    },
    "channel_versions": {
      "NodeA": 4,
      "NodeB": 4,
      "__start__": 2,
      "__start__:NodeA": 3,
      "messages": 4,
    },
    "id": "1ec6aaee-88c3-6804-8002-708967be2a09",
    "pending_sends": [],
    "ts": "2022-01-01T03:00:00.000Z",
    "v": 1,
    "versions_seen": {
      "NodeA": {
        "__start__:NodeA": 2,
      },
      "NodeB": {
        "NodeA": 3,
      },
      "__input__": {},
      "__start__": {
        "__start__": 1,
      },
    },
  },
  "config": {
    "configurable": {
      "checkpoint_id": "1ec6aaee-88c3-6804-8002-708967be2a09",
      "checkpoint_ns": "",
      "thread_id": "1",
    },
  },
  "metadata": {
    "parents": {},
    "source": "loop",
    "step": 2,
    "writes": {
      "NodeB": {
        "messages": [
          AIMessage AIMessage {
            "additional_kwargs": {},
            "content": "Hello from NodeB",
            "id": undefined,
            "invalid_tool_calls": [],
            "lc_kwargs": {
              "additional_kwargs": {},
              "content": "Hello from NodeB",
              "invalid_tool_calls": [],
              "response_metadata": {},
              "tool_calls": [],
            },
            "lc_namespace": [
              "langchain_core",
              "messages",
            ],
            "lc_serializable": true,
            "name": undefined,
            "response_metadata": {},
            "tool_calls": [],
            "usage_metadata": undefined,
          },
        ],
      },
    },
  },
  "parentConfig": {
    "configurable": {
      "checkpoint_id": "1ec6aaee-88c3-6803-8001-2107a262873a",
      "checkpoint_ns": "",
      "thread_id": "1",
    },
  },
  "pendingWrites": [],
}
`;
